std/path.d

```diff
@@ -2438,111 +2439,139 @@ in
                     }
+                    auto anymatch = false;
+                    while (1)
+                    {
+                        const pc2 = pattern[pi];
+                        if (pc2 == ']')
+                            break;
+                        if (!anymatch && (filenameCharCmp!cs(nc, pc2) == 0))
+                            anymatch = true;
+                        ++pi;
+                    }
+                    if (anymatch == not)
+                        return false;
+                    break;
+
+                case '{':
+                    // find end of {} section
+                    auto piRemain = pi;
+                    for (; piRemain < pattern.length
+                             && pattern[piRemain] != '}'; ++piRemain)
+                    {   }
+
+                    if (piRemain < pattern.length)
+                        ++piRemain;
+                    ++pi;
 
-                    if (pi0 == pi)
+                    while (pi < pattern.length)
                     {
-                        if (globMatch!(cs, C)(path[ni..$], pattern[piRemain..$]))
+                        const pi0 = pi;
+                        C pc3 = pattern[pi];
+                        // find end of current alternative
+                        for (; pi < pattern.length && pc3 != '}' && pc3 != ','; ++pi)
                         {
-                            return true;
+                            pc3 = pattern[pi];
                         }
-                        pi++;
-                    }
-                    else
-                    {
-                        if (globMatch!(cs, C)(path[ni..$],
-                                        pattern[pi0..pi-1]
-                                        ~ pattern[piRemain..$]))
+
+                        auto p = path.save;
+                        if (pi0 == pi)
                         {
-                            return true;
+                            if (globMatch!(cs, C)(p, pattern[piRemain..$]))
+                            {
+                                return true;
+                            }
+                            ++pi;
+                        }
+                        else
+                        {
+                            /* Match for:
+                             *   pattern[pi0..pi-1] ~ pattern[piRemain..$]
+                             */
+                            if (pattmp.ptr == null)
+                                // Allocate this only once per function invocation.
+                                // Should do it with malloc/free, but that would make it impure.
+                                pattmp = new C[pattern.length];
+
+                            const len1 = pi - 1 - pi0;
+                            pattmp[0 .. len1] = pattern[pi0 .. pi - 1];
+
+                            const len2 = pattern.length - piRemain;
+                            pattmp[len1 .. len1 + len2] = pattern[piRemain .. $];
```

Comment:
You could extend `globMatch` to accept any random index range as pattern, then you could use [`chain`](http://dlang.org/phobos/std_range.html#.chain) here instead of allocating a temporary.